Add new tier 3 target: `x86_64-unknown-none`

mikeleany opened the pull request:
Adds support for compiling OS kernels or other bare-metal applications for the x86-64 architecture.

Below are details on how this target meets the requirements for tier 3:

> A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

I would be willing to be a target maintainer, though I would appreciate if others volunteered to help with that as well.

> Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

Uses the same naming as the LLVM target, and the same convention as many other bare-metal targets.

> Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.

I don't believe there is any ambiguity here.

> Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.

I don't see any legal issues here.

> The target must not introduce license incompatibilities.
> Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).
> The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.
>If the target supports building host tools (such as rustc or cargo), those host tools must not depend on proprietary (non-FOSS) libraries, other than ordinary runtime libraries supplied by the platform and commonly used by other binaries built for the target. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.
> Targets should not require proprietary (non-FOSS) components to link a functional binary or library.
> "onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

I see no issues with any of the above.

> Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.
> This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

Only relevant to those making approval decisions.

> Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

`core` and `alloc` can be used. `std` cannot be used as this is a bare-metal target.

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running tests (even if they do not pass), the documentation must explain how to run tests for the target, using emulation if possible or dedicated hardware if necessary.

Use `--target=x86_64-unknown-none-elf` option to cross compile, just like any target. The target does not support running tests.

> Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.
> Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

I don't foresee this being a problem.

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.
> In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.

No other targets should be affected by the pull request.

mikeleany committed:
Add new target: `x86_64-unknown-none-elf`

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @cjgillot (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-highfive commented:
:warning: **Warning** :warning:

* These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3638990817) [(plain)](https://github.com/rust-lang/rust/commit/9a8b2a57af150b1ff2835743d8f425b3173aa021/checks/3638990817/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs at line 22:
     Target {
         llvm_target: "x86_64-unknown-none-elf".to_string(),
         pointer_width: 64,
-        data_layout: "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128".to_string(),
+        data_layout: "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+            .to_string(),
         arch: "x86_64".to_string(),
         options: opts,
     }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/mipsisa32r6el_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/riscv32i_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/sparc64_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/x86_64_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/android_base.rs" "/checkout/compiler/rustc_target/src/spec/arm_unknown_linux_musleabihf.rs" "/checkout/compiler/rustc_target/src/spec/aarch64_fuchsia.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

mikeleany committed:
Fix code formatting.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3640671588) [(plain)](https://github.com/rust-lang/rust/commit/d76b3e5bbc6c152961d187ec3da1bb84b66e25c4/checks/3640671588/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
platform support check
   Compiling tier-check v0.1.0 (/checkout/src/tools/tier-check)
    Finished release [optimized] target(s) in 0.76s
     Running `build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/tier-check /checkout/src/doc/rustc/src/platform-support.md /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc`
error: target `x86_64-unknown-none-elf` is missing from platform-support.md
If this is a new target, please add it to /checkout/src/doc/rustc/src/platform-support.md.

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "run" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/tier-check/Cargo.toml" "/checkout/src/doc/rustc/src/platform-support.md" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc"
expected success, got: exit status: 1

```

</details>

mikeleany committed:
Add `x86_64-unknown-none-elf` to `platform-support.md`.

joshtriplett commented:
Thank you for working on this. I have a proposal in progress myself for a similar target, and there are a couple of changes I'd propose; I also have target documentation for the target. Let me get that together, please, and I'll post PR review within a day or two.

joshtriplett commented:
(One such change: I believe this should use CodeModel::Kernel.)

mikeleany committed:
Use `CodeModel::Kernel` for `x86_64-unknown-none-elf`.

mikeleany commented:
> Thank you for working on this. I have a proposal in progress myself for a similar target, and there are a couple of changes I'd propose; I also have target documentation for the target. Let me get that together, please, and I'll post PR review within a day or two.

Thanks I appreciate the help.

> (One such change: I believe this should use CodeModel::Kernel.)

Yes, I agree. I've noticed that some JSON files out there have that and others don't. I had intended to include it here, but somehow missed it. That is corrected now.

bjorn3 commented on compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs:
```diff
@@ -0,0 +1,31 @@
+// `target-cpu` compiler flags to opt-in more hardware-specific
+// features.
+//
+// For example, `-C target-cpu=cortex-a53`.
```
This comment needs to be updated.

joshtriplett commented on compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs:
```diff
@@ -0,0 +1,31 @@
+
+pub fn target() -> Target {
+    let opts = TargetOptions {
+        abi: "softfloat".to_string(),
```
This shouldn't be here. `softfloat` is an ARM ABI, not an x86 one. The target just shouldn't have an `ABI` field.

joshtriplett commented on compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs:
```diff
@@ -0,0 +1,31 @@
+        abi: "softfloat".to_string(),
+        linker_flavor: LinkerFlavor::Lld(LldFlavor::Ld),
+        linker: Some("rust-lld".to_owned()),
+        features: "-mmx,-sse,+soft-float".to_string(),
```
```suggestion
        features: "-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2,+soft-float".to_string(),
```

(Taken from `x86_64-unknown-none-linuxkernel`.)

joshtriplett commented on compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs:
```diff
@@ -0,0 +1,31 @@
+use super::{CodeModel, LinkerFlavor, LldFlavor, PanicStrategy, Target, TargetOptions};
+
+pub fn target() -> Target {
+    let opts = TargetOptions {
```
```suggestion
    let opts = TargetOptions {
        cpu: "x86-64".to_string(),
        max_atomic_width: Some(64),
```

joshtriplett committed:
Rename x86_64-unknown-none-elf to x86_64-unknown-none

Most Rust freestanding/bare-metal targets use just `-unknown-none` here,
including aarch64-unknown-none, mipsel-unknown-none, and the BPF
targets. The *only* target using `-unknown-none-elf` is RISC-V.

The underlying toolchain doesn't care; LLVM accepts both `x86_64-unknown-none`
and `x86_64-unknown-none-elf`.

In addition, there's a long history of embedded x86 targets with varying
definitions for the `elf` suffix; on some of those embedded targets,
`elf` implied the inclusion of a C library based on newlib or similar.
Using `x86_64-unknown-none` avoids any potential ambiguity there.

(Work on this target sponsored by Profian.)

joshtriplett committed:
x86_64-unknown-none: Fix module comment

(Work on this target sponsored by Profian.)

joshtriplett committed:
x86_64-unknown-none: Drop the `abi` field

(Work on this target sponsored by Profian.)

joshtriplett committed:
x86_64-unknown-none: Disable more target features

Based on the list used for x86_64-unknown-none-linuxkernel.

(Work on this target sponsored by Profian.)

joshtriplett committed:
x86_64-unknown-none: Expand TargetOptions to specify more details

Specify the `cpu` and the `max_atomic_width` (64).

Set `stack_probes` similarly to other targets to work around known
issues, and copy the corresponding comment from those targets.

Build position-independent code that doesn't require relocations.

(Work on this target sponsored by Profian.)

Aaron1011 commented on compiler/rustc_target/src/spec/x86_64_unknown_none_elf.rs:
```diff
@@ -0,0 +1,31 @@
+// Generic AArch64 target for bare-metal code - Floating point disabled
```
```suggestion
// Generic X86 target for bare-metal code - Floating point disabled
```

joshtriplett committed:
x86_64-unknown-none: Add target documentation

In particular, document the default properties and assumptions of code
built for the target.

(Work on this target sponsored by Profian.)

joshtriplett commented:
I sent a PR to this PR, at https://github.com/mikeleany/rust/pull/1 with additions and modifications to this proposed target. This also addresses all of the outstanding feedback (including my own), and adds target-specific documentation.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3646347439) [(plain)](https://github.com/rust-lang/rust/commit/88d4df648eb1a2c2d0d6047e6d702d72cd18be19/checks/3646347439/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_parse_format v0.0.0 (/checkout/compiler/rustc_parse_format)
error[E0433]: failed to resolve: use of undeclared type `StackProbeType`
  --> compiler/rustc_target/src/spec/x86_64_unknown_none.rs:14:23
   |
14 |         stack_probes: StackProbeType::Call,
   |                       ^^^^^^^^^^^^^^ use of undeclared type `StackProbeType`

error[E0433]: failed to resolve: use of undeclared type `RelroLevel`
  --> compiler/rustc_target/src/spec/x86_64_unknown_none.rs:17:22
   |
17 |         relro_level: RelroLevel::Full,
   |                      ^^^^^^^^^^ use of undeclared type `RelroLevel`
error[E0433]: failed to resolve: use of undeclared type `RelocModel`
  --> compiler/rustc_target/src/spec/x86_64_unknown_none.rs:18:27
   |
18 |         relocation_model: RelocModel::Static,
```

</details>

mikeleany commented:
> I sent a PR to this PR, at [mikeleany#1](https://github.com/mikeleany/rust/pull/1) with additions and modifications to this proposed target. This also addresses all of the outstanding feedback (including my own), and adds target-specific documentation.
> 
> I also submitted the target tier request as a compiler MCP at [rust-lang/compiler-team#462](https://github.com/rust-lang/compiler-team/issues/462) . That proposal adds a second target maintainer, @haraldh.

Thanks for this and everyone else's feedback.

I've removed -elf from the PR title to reflect that change. I had debated whether that should be included or not. I can see the point that it may imply a level of library support.

joshtriplett commented:
I'll fix the compilation errors; need to add imports for the appropriate enums.

mikeleany committed:
Fix build errors.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3646584994) [(plain)](https://github.com/rust-lang/rust/commit/c02a30bf2ecb11f2d0de6365638dc95961356b4f/checks/3646584994/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({"src/doc/edition-guide"}) not skipped for "bootstrap::doc::EditionGuide" -- not in ["src/tools/tidy"]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
rustc/print.html:6766: broken link - `rustc/platform-support/x86_64-unknown-none.html`
rustc/platform-support.html:392: broken link - `rustc/platform-support/x86_64-unknown-none.html`
number of HTML files scanned: 28337
number of HTML redirects found: 8809
number of links checked: 1774729
number of links ignored due to external: 66217
```

</details>

mikeleany committed:
Documentation fixes.

Fix broken link to target documentation.

Also fix formatting of developer list.

joshtriplett commented:
The current failure looks like a bug in the linkchecker with mailto links. We can work around it, but I think the linkchecker may need improvement here.

haraldh commented on compiler/rustc_target/src/spec/x86_64_unknown_none.rs:
```diff
@@ -0,0 +1,41 @@
+        executables: true,
+        disable_redzone: true,
+        panic_strategy: PanicStrategy::Abort,
+        code_model: Some(CodeModel::Kernel),
```
Do we really want `CodeModel::Kernel` ??

haraldh replied:
I would vote for `CodeModel::Large` to be as generic as possible.

joshtriplett commented:
On Mon, Sep 20, 2021 at 12:21:00AM -0700, Harald Hoyer wrote:
> @haraldh commented on this pull request.
> 
> 
> 
> > +        max_atomic_width: Some(64),
> +        // don't use probe-stack=inline-asm until rust#83139 and rust#84667 are resolved
> +        stack_probes: StackProbeType::Call,
> +        position_independent_executables: true,
> +        static_position_independent_executables: true,
> +        relro_level: RelroLevel::Full,
> +        relocation_model: RelocModel::Static,
> +        linker_flavor: LinkerFlavor::Lld(LldFlavor::Ld),
> +        linker: Some("rust-lld".to_owned()),
> +        features:
> +            "-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2,+soft-float"
> +                .to_string(),
> +        executables: true,
> +        disable_redzone: true,
> +        panic_strategy: PanicStrategy::Abort,
> +        code_model: Some(CodeModel::Kernel),
> 
> Do we really want `CodeModel::Kernel` ??

For freestanding/bare-metal usage, I think we do, yes. Code running in
ring 0 most often wants to be using this, especially if it supports ring
3 but often even if it doesn't.

A project can always change the code model, but I think this is the
right default for this target.

jackpot51 commented:
If anyone needs another reference for this, here is the Redox kernel's default target spec. It would be nice to be able to use a built-in one.

```
{
    "llvm-target": "x86_64-unknown-none",
    "target-endian": "little",
    "target-pointer-width": "64",
    "target-c-int-width": "32",
    "data-layout": "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
    "arch": "x86_64",
    "os": "none",
    "env": "",
    "vendor": "unknown",
    "linker-flavor": "gcc",
    "target-family": "redox",
    "pre-link-args": {
        "gcc": ["-m64", "-nostdlib", "-static"]
    },
    "features": "-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2,+soft-float",
    "dynamic-linking": false,
    "executables": false,
    "relocation-model": "pic",
    "code-model": "kernel",
    "disable-redzone": true,
    "eliminate-frame-pointer": false,
    "exe-suffix": "",
    "has-rpath": false,
    "no-compiler-rt": true,
    "no-default-libraries": true,
    "position-independent-executables": false,
    "has-elf-tls": true
}
```

jackpot51 commented:
I defined this target spec above as `x86_64-unknown-kernel`. Perhaps there is value in having `x86_64-unknown-none` use the default code model, but have another target like `x86_64-unknown-kernel` which uses the kernel code model?

joshtriplett commented:
@jackpot51 I don't think code model alone is enough to motivate a completely different target name, given that we already have a stable rustc option to change it. (We should ideally have a stable cargo option for it as well.) I think we should pick a default and then document how to change it. (The target documentation mentions how to change it.)

I also don't think we'd want two different targets for it that Rust code would potentially use `cfg` to distinguish.

jackpot51 commented:
Yeah, good point. I haven't seen a bare metal x86_64 target _not_ using kernel codemodel, are there examples of that?

mikeleany commented:
@jackpot51 
> Yeah, good point. I haven't seen a bare metal x86_64 target _not_ using kernel codemodel, are there examples of that?

Yes, I've seen a few examples in Rust, though I don't know if that was a conscious decision or if the code model just wasn't considered. [Philipp Oppermann's Blog OS](https://os.phil-opp.com/minimal-rust-kernel/#putting-it-together) is one example. I think that target specification has probably been copied verbatim a few times without giving much thought to what's in it (including by me).

The osdev wiki also has an [article](https://wiki.osdev.org/Building_libgcc_for_mcmodel%3Dkernel) which suggests using either kernel or large, which would imply that there may be at least a few projects out there where someone intentionally chose a large code model over kernel.

However, kernel should be a sane default, and if people want to change it they can.

bjorn3 commented:
According to https://wiki.osdev.org/X86-64 the kernel code model requires the kernel to run in the upper 2GB of the memory. Blog OS runs near the start of the memory and as such won't work without UB with the kernel code model.

haraldh commented:
> According to https://wiki.osdev.org/X86-64 the kernel code model requires the kernel to run in the upper 2GB of the memory. Blog OS runs near the start of the memory and as such won't work without UB with the kernel code model.

https://stackoverflow.com/questions/49381147/use-of-mcmodel-kernel-flag-in-x86-platform

AFAIU, In the case of the lower 2GB, this is covered by the kernel code model.

small code model: `mov eax, imm32`
kernel code model: `mov rax, sign_extended_imm32`

haraldh commented:
Also found this nice blog post:
https://eli.thegreenplace.net/2012/01/03/understanding-the-x64-code-models/

joshtriplett committed:
Unlinkify email address to satisfy linkchecker

The linkchecker doesn't seem happy with links to email addresses.

joshtriplett commented:
PR to work around the link checking: https://github.com/mikeleany/rust/pull/2

(We shouldn't block on analyzing whether or not there's a linkchecker issue here.)

bors commented:
:umbrella: The latest upstream changes (presumably #86191) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
I think the open issues around this have been settled at this point, other than confirming the naming. Between discussions here and on Zulip, several folks seem to agree that `-elf` may imply some level of library support that we don't want to imply; however, I'm not sure if there's full consensus on that. Existing Rust targets aren't consistent on this front, with some using `-none` and some using `-none-elf`. I personally would favor `-none`, but I also don't think this is something we should block on.

I think people would get used to either name, and there'd be value in picking one and going forward with it.

Would there be objections to proceeding with `-none`?

npmccallum commented on compiler/rustc_target/src/spec/x86_64_unknown_none.rs:
```diff
@@ -0,0 +1,41 @@
+        position_independent_executables: true,
+        static_position_independent_executables: true,
+        relro_level: RelroLevel::Full,
+        relocation_model: RelocModel::Static,
```
@haraldh We are currently using `relocation-model=pic` in our code. But I don't remember why. Does this represent a problem for us?

joshtriplett replied:
I thought that we had changed this; I absolutely think we should set pic as the default.

joshtriplett replied:
Fixed in https://github.com/mikeleany/rust/pull/3

haraldh replied:
> @haraldh We are currently using `relocation-model=pic` in our code. But I don't remember why. Does this represent a problem for us?

@npmccallum thx for spotting this

joshtriplett committed:
x86_64-unknown-none: Use position-independent code by default

This avoids requiring relocation code, which a bare-metal environment
may not have or want.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3762797239) [(plain)](https://github.com/rust-lang/rust/commit/d76204bf3668e3c1ca112f34a1f28f21048a12e9/checks/3762797239/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Suite("src/test/codegen") not skipped for "bootstrap::test::Codegen" -- not in ["src/tools/tidy"]
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 293 tests
ii.....iiiF..i...i..i.......iii........ii.....i.iii..........i.ii..............i.i..............i... 100/293
Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
........................i.ii......ii...i...........................iiiii.i...................
failures:


---- [codegen] codegen/abi-efiapi.rs#aarch64 stdout ----

error in revision `aarch64`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-10/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/abi-efiapi.aarch64/abi-efiapi.ll" "/checkout/src/test/codegen/abi-efiapi.rs" "--check-prefixes" "CHECK,NONMSVC,aarch64"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
/checkout/src/test/codegen/abi-efiapi.rs:29:12: error: aarch64: expected string not found in input
//aarch64: define dso_local void @has_efiapi
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/abi-efiapi.aarch64/abi-efiapi.ll:1:1: note: scanning from here
; ModuleID = 'abi_efiapi.1d5bfd15-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/abi-efiapi.aarch64/abi-efiapi.ll:7:1: note: possible intended match here
define void @has_efiapi() unnamed_addr #0 {

------------------------------------------


---
test result: FAILED. 229 passed; 1 failed; 63 ignored; 0 measured; 0 filtered out; finished in 2.75s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "codegen" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:10:51
```

</details>

npmccallum commented:
> Would there be objections to proceeding with `-none`?

None from me.

npmccallum commented:
Here's my thoughts about code model (relevant to our project). In Enarx we have two binaries that want to use this target. One is for Intel SGX and the other is a VM kernel.

From the perspective of our VM kernel, I think we can use any of the proposed code models because we have full control over layout. Also the kernel is small. So `kernel` probably makes the most sense here.

The SGX case is more tricky. The size of the binary is very small. So no problem there. However, in this case the code is loaded into an SGX enclave. The enclave (the total enclave space, not merely the binary size) is currently 2GiB in size and must be naturally aligned. Therefore, it can't be in the lowest 2GiB. And because this is userspace, it can't be in the highest 2GiB either. Further, we currently allow the kernel (via mmap) to choose the location of the enclave.

It isn't at all clear to me what happens if we build with `mcmodel=kernel` for SGX. I've tried it and everything seems to work. But I don't know what dragons are hidden in the output. Does anyone have a good sense of precisely what the impacts of this would be?

haraldh commented:
It's about symbols pointing to somewhere. If those symbols in the binary be it data or code point to locations >2G then we need bigger models. "medium" if data symbols are located far away. "large" if code symbols can be far away.

haraldh commented:
Personally I would trade in a larger binary for universal use cases.

npmccallum commented:
@haraldh Yes, that part I understand. What isn't clear to me is the failure mode. What happens if you build a PIE with `mcmodel=kernel` but then load it at 1TiB? Does this result in UB? If so, that could be spectacularly bad (particularly from a security perspective).

I don't understand where the "lower 2GiB" or "upper 2GiB" requirement comes from. It seems to me that this is mostly about the binary size itself: the maximum possible distance from RIP in the PIC case. This distance should always be a signed 32-bit offset from RIP. However, this criteria can be met so long as all linking is done within a 2GiB window (i.e. the maximum distance between any two addresses is <2GiB). Therefore, I don't understand why the "lower/upper 2GiB" requirement arises.

bjorn3 commented:
https://stackoverflow.com/questions/49381147/use-of-mcmodel-kernel-flag-in-x86-platform

> If you didn't get link errors without -mcmodel=kernel (like these), you probably have a gcc that makes PIE executables by default (common on recent distros), so it avoids absolute addressing.

This seems to suggest that for PIE this restriction of the kernel code model doesn't exist.

npmccallum commented:
@bjorn3 For my clarification, I hear you saying:
1. `mcmodel=kernel` can be loaded anywhere if only PIC code is emitted
2. for PIC-only code, `small` and `kernel` are identical
3. the link location requirements only apply to absolute addressing (i.e. non-PIC)
4. failures are detected at link time

Is that correct?

npmccallum commented:
# Differences with x86_64-unknown-none-linuxkernel

I thought that it might be helpful to highlight the differences between this proposed target and the `linuxkernel` target. This has two goals:
1. Check if we've missed anything we want.
2. Check if merging the targets is possible.

Therefore, I've manually compiled a list of differences expressed in a pseudo-unified-diff format as if we were diffing from this proposal to `linuxkernel`:

```
- stack_probes: StackProbeType::Call,
+ stack_probes: StackProbeType::InlineOrCall { min_llvm_version_for_inline: (11, 0, 1) },
- executables: true,
- static_position_independent_executables: true,
+ eliminate_frame_pointer: false,
+ needs_plt: true,
- relocation_model: RelocModel::Pic,
+ relocation_model: RelocModel::Static,
- linker_flavor: LinkerFlavor::Lld(LldFlavor::Ld),
- linker: Some("rust-lld".to_owned()),
+ env: “gnu”
+ linker_is_gnu: true,
+ pre_link_args: (LinkerFlavor::Gcc, ["-Wl,--as-needed", "-Wl,-z,noexecstack", "-m64"])
```

## Stack Probes

```
- stack_probes: StackProbeType::Call,
+ stack_probes: StackProbeType::InlineOrCall { min_llvm_version_for_inline: (11, 0, 1) },
```

If I understand this correctly, the `linuxkernel` option is the better one since it allows for inline stack probes which are an optimization over calling `__rust_probestack`.

Proposal: Change `x86_64-unknown-none` target to match `x86_64-unknown-none-linuxkernel`.

## Executable Type: Static vs PIC

```
  position_independent_executables: true,
- executables: true,
- static_position_independent_executables: true,
- relocation_model: RelocModel::Pic,
+ relocation_model: RelocModel::Static,
```

Both targets claim to allow `position_independent_executables`, but `linuxkernel` disables PIC. If `linuxkernel` can work with PIC we should be able to combine all these options.

Proposal: Get input from `linuxkernel` maintainers (@alex ?)

## Linker: GNU vs LLD

```
- linker_flavor: LinkerFlavor::Lld(LldFlavor::Ld),
- linker: Some("rust-lld".to_owned()),
+ linker_is_gnu: true,
```

This might be a deal-breaker for merging the targets. Does `linuxkernel` really require the GNU linker? If so, how horrible is it if we do so too? Is there a reason we require LLD?

Are the consumers of `linuxkernel` using `cargo`? If they are calling `rustc` directly, they can always do `-C linker=...`. However, changing the linker from `cargo` is currently difficult.

Proposal: Get input from `linuxkernel` maintainers (@alex ?) and @mikeleany / @haraldh 

## Linker Arguments

```
+ pre_link_args: (LinkerFlavor::Gcc, ["-Wl,--as-needed", "-Wl,-z,noexecstack", "-m64"])
```

I'm fairly certain these linker arguments are not required here and could be implemented using cargo or rustc directly.

Proposal: If merging with `linuxkernel`, drop these arguments.

## Frame Pointers

```
+ eliminate_frame_pointer: false,
```

This seems like a pretty common thing to turn on for kernels. The cost for doing so is pretty low.

Proposal: Change `x86_64-unknown-none` target to keep frame pointers.

## Miscellaneous

```
+ needs_plt: true,
+ env: “gnu”
```

I don't know why the Linux kernel wants these. I also don't know if we want them or not.

Proposal: Get input from `linuxkernel` maintainers (@alex ?) and @mikeleany / @haraldh

npmccallum commented:
# Differences with x86_64-unknown-none-hermitkernel

Most of the differences between this platform and `hermitkernel` are similar to those in `linuxkernel` and they could likely be combined. However, one big difference is apparent between the targets:

```
+ has_elf_tls: true,
+ tls_model: TlsModel::InitialExec,
```

@stlankes Is thread-local storage required for HermitCore? Is there some way we could coalesce these requirements?

bjorn3 commented:
> This might be a deal-breaker for merging the targets. Does linuxkernel really require the GNU linker? If so, how horrible is it if we do so too? Is there a reason we require LLD?

This is just the default. You can explicitly override the linker.

> Is there a reason we require LLD?

The default for baremetal targets should be lld IMHO as it avoids the need to install a linker capable of linking for the respective target. We should avoid a linker wrapper like gcc as it may pass extra options on some OSes that are incompatible with baremetal targets.

> I'm fairly certain these linker arguments are not required here and could be implemented using cargo or rustc directly.

This **is** direcrlt implemented using rustc. The target spec is part of rustc.

> I don't know why the Linux kernel wants these. I also don't know if we want them or not.

`needs_plt` is for kernel modules I would guess. `env` doesn't make sense to me.

npmccallum commented:
> > This might be a deal-breaker for merging the targets. Does linuxkernel really require the GNU linker? If so, how horrible is it if we do so too? Is there a reason we require LLD?
> 
> This is just the default. You can explicitly override the linker.

Sort of.

1. You can ask whoever is building the software to pass `-C linker=...` to `rustc` manually. This is fine if you're invoking `rustc` in a makefile. However, you can't do this during a `cargo build`; so this isn't user friendly to consumers of the software who don't understand how these options work.

2. You can hard code the linker in `.cargo/config`. This is more user friendly to projects using cargo and who want the simple `cargo build` experience. However, there are drawbacks. First, this only works when you want the same linker on all supported targets and linker versions. This is because there is no dynamic way to set the linker. Second, the use of `.cargo/config` is broadly incompatible work cargo workspaces. So this breaks in common cases where you want to build multiple binary crates in a workspace.

3. There is currently no way to specify the linker from a build script. This seems, to me, like an obvious build script option. But it wasn't received well when I suggested it in the past. There is no timeline to implement this feature.

> > Is there a reason we require LLD?
> 
> The default for baremetal targets should be lld IMHO as it avoids the need to install a linker capable of linking for the respective target. We should avoid a linker wrapper like gcc as it may pass extra options on some OSes that are incompatible with baremetal targets.

I misunderstood the details of these linker arguments. `linker_is_gnu` only defines whether the linker understands GNU-like arguments, which LLD does. So it looks like `x86_64-unknown-none-linuxkernel` is choosing the configured default linker (i.e. from `TargetOptions::default()` but asserting that it is GNU-like. OTOH, `x86_64-unknown-none` is forcing the default target linker to LLD.

It seems to me like we can find a compromise here.

> > I'm fairly certain these linker arguments are not required here and could be implemented using cargo or rustc directly.
> 
> This **is** direcrlt implemented using rustc. The target spec is part of rustc.

Sorry, I wasn't clear. By "... implemented using cargo or rustc directly" I meant something different. `linuxkernel` is requesting three linker arguments: `["-Wl,--as-needed", "-Wl,-z,noexecstack", "-m64"]`. I suspect that `-m64` is redundant and can be dropped. But if it isn't, we can clearly add it to our target.

OTOH, `-Wl,--as-needed` and `-Wl,-z,noexecstack` don't really seem to me to be **target** linker flags so much as application linker flags. Therefore, they can be moved from the target specification into either:

1. A cargo build script using [extra-link-arg](https://doc.rust-lang.org/cargo/reference/unstable.html#extra-link-arg).

2. Or, if the Linux kernel is using `make` to invoke `rustc` directly, the makefiles can be updated to add `-C link-arg=...`.

stlankes commented:
> Is thread-local storage required for HermitCore? Is there some way we could coalesce these requirements?

@npmccallum HermitCore is library operating system. Consequently, I link the kernel to a common application. The application is able to use TLS. But I think that the feature isn't required for the kernel. I will check it.

bjorn3 commented:
`-Wl,--as-needed` is used on other targets too to avoid DT_NEEDED for dynamic libraries that don't provide any required symbols. We actually define it as being the default I believe. There is a linler modifier to disable it for specific libraries, but the code to handle this works incorrectly when not globallt using as-needed I think. `-Wl,-z,noexecstack` is a security mitigation and as such also in scope for being passed by default by rustc. -z,execstack only exists because of the gcc scoped functions support for C. Rust doesn't need it, nor does portable C.

npmccallum commented:
> `-Wl,--as-needed` is used on other targets too to avoid DT_NEEDED for dynamic libraries that don't provide any required symbols. We actually define it as being the default I believe. There is a linler modifier to disable it for specific libraries, but the code to handle this works incorrectly when not globallt using as-needed I think. `-Wl,-z,noexecstack` is a security mitigation and as such also in scope for being passed by default by rustc. -z,execstack only exists because of the gcc scoped functions support for C. Rust doesn't need it, nor does portable C.

Yes. However, while it is trivial to **add** a linker flag using either `build.rs` or `rustc -C link-arg...` it is currently impossible to **remove** or **reorder** a linker flag specified by the target specification. So to create a target specification that has maximum usefulness, we should not specify any linker flags that are not absolutely required.

bjorn3 commented:
As I said you can use a linker modifier to disable as-needed for specific libraries. If you really want you can also use `-Wl,--no-as-needed`, which will override the `-Wl,--as-needed` though that could still cause issues around the as-needed linker modifier handling (think forgetting a `-Wl,--(no-)as-needed` required to match the linker modifier as it expected that the current state was different) just like not passing `-Wl,--as-needed` in the first place would.

https://rust-lang.github.io/rfcs/2951-native-link-modifiers.html#as-needed

joshtriplett commented:
Given that the target defaults to PIC, as far as I can tell, it doesn't matter where precisely you load the binary, as long as all the code and data are within 2GB of each other. The code model only comes into play if you have references to things more than 2GB away, which typically means you have data or code that's more than 2GB in size.

Also, as far as I can tell, almost all existing Rust targets with a code model default to whatever LLVM defaults to, and LLVM defaults to `small`.

Given all of that, I don't think we need to use `large` by default; that seems like overkill. And it's not just that doing so would make binaries larger; it would also make memory accesses slower.

Any application that cares about the code model would need to either be disabling PIC, or having code and data larger than 2GB. In the former case, the application is already changing options so it can change the code model as well. The latter case is unusual, and seems likely to have other special requirements. I think it'd be reasonable for both cases to need to change options itself.

Also, as far as I can tell, if you use a code model that's too small for your application, you won't get silent failures; you'll get a linker error.

bors commented:
:umbrella: The latest upstream changes (presumably #88952) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
@npmccallum I agree with that summary; I don't think there are any outstanding unresolved questions at this point.

@rust-lang/compiler, I think this is ready for a decision (either here or in https://github.com/rust-lang/compiler-team/issues/462). (That can happen in parallel with the PR being rebased to eliminate the conflict.)

pthariensflame reviewed (changes requested):
It appears that the PIC fix PR “fixed” the wrong target spec!

mikeleany committed:
Fix issue where PIC was added to the wrong target.

Should be for x86_64_unknown_none, but aarch64_unknown_none was inadvertently
updated instead.

stlankes commented:
@npmccallum I removed TLS support  in x86_64-unknown-none-hermitkernel. It isn't necessary.

joshtriplett commented:
Current status: the compiler team MCP has been fully approved, and all the comments on this PR have been addressed, so someone needs to do the final r+ on it. I'll track down someone to do so.

cjgillot commented:
The MCP has been accepted, and the feedback appears to have been addressed. This is Tier 3, and there does not appear to be an impact on other targets.
r? @cjgillot 
@bors r+

bors commented:
:pushpin: Commit 11d54dc1c61f74aa4cfa704ed97bf6c853a731e5 has been approved by `cjgillot`

<!-- @bors r=cjgillot 11d54dc1c61f74aa4cfa704ed97bf6c853a731e5 -->
<!-- homu: {"type":"Approved","sha":"11d54dc1c61f74aa4cfa704ed97bf6c853a731e5","approver":"cjgillot"} -->

bors commented:
:hourglass: Testing commit 11d54dc1c61f74aa4cfa704ed97bf6c853a731e5 with merge ff0e14829e1806ca0d4226595f7fdf3e8658758f...
<!-- homu: {"type":"BuildStarted","head_sha":"11d54dc1c61f74aa4cfa704ed97bf6c853a731e5","merge_sha":"ff0e14829e1806ca0d4226595f7fdf3e8658758f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4062157792?check_suite_focus=true)
Approved by: cjgillot
Pushing ff0e14829e1806ca0d4226595f7fdf3e8658758f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4062157792?check_suite_focus=true"},"merge_sha":"ff0e14829e1806ca0d4226595f7fdf3e8658758f"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#89062!

Tested on commit rust-lang/rust@ff0e14829e1806ca0d4226595f7fdf3e8658758f.
Direct link to PR: <https://github.com/rust-lang/rust/pull/89062>

💔 rls on linux: test-pass → test-fail (cc @Xanewok).

rust-timer commented:
Finished benchmarking commit (ff0e14829e1806ca0d4226595f7fdf3e8658758f): [comparison url](https://perf.rust-lang.org/compare.html?start=c7e4740ec18996e082fe6e29ebf7efdc7dda418f&end=ff0e14829e1806ca0d4226595f7fdf3e8658758f).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

haraldh commented:
> @npmccallum I removed TLS support in x86_64-unknown-none-hermitkernel (see #89941). It isn't necessary.

so, I tried rusty_demo with:
```diff
❯ git diff
diff --git a/hermit-sys/build.rs b/hermit-sys/build.rs
index 391fae2..3be0e47 100644
--- a/hermit-sys/build.rs
+++ b/hermit-sys/build.rs
@@ -35,7 +35,7 @@ fn build_hermit(src_dir: &Path, target_dir_opt: Option<&Path>) {
                        .arg("-Z")
                        .arg("build-std=core,alloc")
                        .arg("--target")
-                       .arg("x86_64-unknown-none-hermitkernel");
+                       .arg("x86_64-unknown-none");
        } else if target_arch == "aarch64" {
                cmd.current_dir(src_dir)
                        .arg("build")
@@ -132,7 +132,7 @@ fn build_hermit(src_dir: &Path, target_dir_opt: Option<&Path>) {
 
        let lib_location = if target_arch == "x86_64" {
                target_dir
-                       .join("x86_64-unknown-none-hermitkernel")
+                       .join("x86_64-unknown-none")
                        .join(&profile)
                        .canonicalize()
                        .unwrap() // Must exist after building

```

and it seems to work. Any blockers on using `x86_64-unknown-none`?

haraldh commented:
@stlankes see above

stlankes commented:
@haraldh I will check it, but I don't see any blocker. But a double check will be great. @mkroening Do you see any issue?

stlankes commented:
@haraldh I thought a little bit about your request. The main issue is that our llvm target is also defined in the spec file (see https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/x86_64_unknown_none_hermitkernel.rs#L14). This is used to define our kernel as standalone kernel (https://github.com/llvm/llvm-project/blob/e356027016c6365b3d8924f54c33e2c63d931492/llvm/include/llvm/MC/MCELFObjectWriter.h#L76-L77). In our case, we develop a library operating systems and everything is linked to one single boot image. Maybe it is currently working, but if I understand everything correctly, every part of the image should use the same OSABI. What do you think?

stlankes commented:
By the way, my Linux defines the OSABI (see /usr/include/elf.h) as following: 

> #define EI_OSABI        7               /* OS ABI identification */
#define ELFOSABI_NONE           0       /* UNIX System V ABI */
#define ELFOSABI_SYSV           0       /* Alias.  */
#define ELFOSABI_HPUX           1       /* HP-UX */
#define ELFOSABI_NETBSD         2       /* NetBSD.  */
#define ELFOSABI_GNU            3       /* Object uses GNU ELF extensions.  */
#define ELFOSABI_LINUX          ELFOSABI_GNU /* Compatibility alias.  */
#define ELFOSABI_SOLARIS        6       /* Sun Solaris.  */
#define ELFOSABI_AIX            7       /* IBM AIX.  */
#define ELFOSABI_IRIX           8       /* SGI Irix.  */
#define ELFOSABI_FREEBSD        9       /* FreeBSD.  */
#define ELFOSABI_TRU64          10      /* Compaq TRU64 UNIX.  */
#define ELFOSABI_MODESTO        11      /* Novell Modesto.  */
#define ELFOSABI_OPENBSD        12      /* OpenBSD.  */
#define ELFOSABI_ARM_AEABI      64      /* ARM EABI */
#define ELFOSABI_ARM            97      /* ARM */
#define ELFOSABI_STANDALONE     255     /* Standalone (embedded) application */

haraldh commented:
Moved this discussion here: https://github.com/hermitcore/rusty-hermit/issues/197
